### ISSUE TITLE
Why are ITRB instances slow as snails?

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#2093]
repro:
1) Go to https://arax.ncats.io/beta/ and issue the example JSON query
2) Examine the first and last timestamp in the streamed progress
3) Note that it is 2-3 seconds between first and last message!

4) Go to https://arax.ci.transltr.io/ and issue the example JSON query
5) Examine the first and last timestamp in the streamed progress
6) Note that it takes 13 seconds!

Tasks:
1. compare the reported progress timestamps to figure out what's taking so long
2. Fix it!

**Comment by isbluis** [comment]
First clue:  Expand query to rtx-kg2 takes ~7 seconds in ITRB (top) vs. <1 second in beta (bottom) :

![image](https://github.com/RTXteam/RTX/assets/37852256/de215c10-482e-4aac-8d5a-1a5c6fab4939)

---

![image](https://github.com/RTXteam/RTX/assets/37852256/ba7a8ff5-45e0-4780-9451-eab31096e8c5)

**Comment by isbluis** [comment]
Another >1.7 seconds are spent building the Message.

CI takes 1.75 secs:
![image](https://github.com/RTXteam/RTX/assets/37852256/23234343-2f15-4de5-a729-2885ad435b99)

---

beta takes 0.03 secs:
![image](https://github.com/RTXteam/RTX/assets/37852256/7c88dfde-11ea-4e8e-953e-07d334136042)

**Comment by isbluis** [comment]
The other significant time delay is in storing the response Messages in MySQL and S3.

CI takes a total of ~1.34 secs:
![image](https://github.com/RTXteam/RTX/assets/37852256/f00c761c-daf3-479b-9305-25345a09f4c8)

---

beta takes ~0.27 secs:
![image](https://github.com/RTXteam/RTX/assets/37852256/98b14e2c-b433-4efd-95e8-da0baad12d74)

**Comment by saramsey** [comment]
Is it possible that this is because `kg2cplover.rtx.ai` is running in `us-west-2` but the ITRB instances are (I think) all running in `us-east-1`?

**Comment by edeutsch** [comment]
okay, thanks, it looks like 3 anomalies are identified. let's call them ABC and work them separately:
A) Slower KG2 request
B) Slower Message building
C) Slower S3 storage

**Comment by edeutsch** [comment]
Regarding A) Slower KG2 request

I don't think this would be the problem. I'm not certain of the current state of things but I am under the impression that both arax.ci.transltr.io and arax.ncats.io/beta should be using the ITRB CI PloverDB instance. So, if anything, ITRB should be FASTER. We should try to trace that.

And in any case, a difference in data center shouldn't lead to a 6 second difference! Maybe a 0.6 second difference..

**Comment by edeutsch** [comment]
Regarding B) Slower Message building

This should be traced and fixed. No good reason for this that I can think of

**Comment by edeutsch** [comment]
C) Slower S3 storage

Well, this 0.7 seconds is our penalty for being in different data centers, I guess. I'm guessing the S3 bucket is in west and ITRB is in east. 

We could consider moving our S3 bucket to east. Then ITRB would be faster and our own arax.ncats.io would pay the penalty. That would probably be a good move. 

but I think this is the most complex fix for the least gain, so I would leave this for last, if at all.

**Comment by edeutsch** [comment]
> Is it possible that this is because `kg2cplover.rtx.ai` is running in `us-west-2` but the ITRB instances are (I think) all running in `us-east-1`?

In case my answer was lost in the various above, I believe the answer is no. Primarily because we are no longer using kg2cplover.rtx.ai. I think it can even be shut down. I'm not 100% certain.

**Comment by edeutsch** [comment]
It just occurs to me that problem A might be due to problem B. If we solve B, that might automatically resolve A.

**Comment by edeutsch** [comment]
Neither of those do much themselves. Should be lightning quick, but they do call other things that might not be fast (but should be). Messenger calls this:
```
        biolink_helper = BiolinkHelper()
```
I have no idea what lies down that well. This is highly reminiscent of issue #1998 
I wonder if booting up up BiolinkHelper takes several seconds on ITRB nodes for some reason?

**Comment by isbluis** [comment]
A 4th anomaly, which is easier to distinguish using a newly-implemented UI feature:

D) Slower Response building

CI takes ~1.6 seconds:
![image](https://github.com/RTXteam/RTX/assets/37852256/7e69f9db-d4fd-4e82-b50d-7d67f09d959d)

---

beta takes under 0.1 seconds:
![image](https://github.com/RTXteam/RTX/assets/37852256/31da52d5-920d-4d48-9077-586166cdd882)


BiolinkHelper may also be the culprit...?

**Comment by isbluis** [comment]
Worth noting that when I run this on the command line in my dev area:
`time python3 ARAX/BiolinkHelper/biolink_helper.py`

It takes almost 2 seconds, most of which are spent building the lookup map:
![image](https://github.com/RTXteam/RTX/assets/37852256/813e025e-f6f4-44bf-825f-4ff57e4e87fb)

Could it be that the CI instances do not have a saved map at the expected location?
`self.biolink_lookup_map_path = f"{biolink_helper_dir}/biolink_lookup_map_{self.biolink_version}_v4.pickle"`

The INFO line would be in the application/web logs -- do we have access to those?  Or how can we verify that this file exists?

**In short: we get a ~2 second hit every time BiolinkHelper is instantiated if that pickle file is not found.**

**Comment by edeutsch** [comment]
Yes, you can see the logs on any instance with something like:
https://kg2.ci.transltr.io/api/rtxkg2/v1.4/status/logs
(currently okay in a browser window but can get unwieldy when running for a long period)

I do not see "INFO: Build" anywhere in that log currently.
Is that code pushed to CI?
Writing to STDERR?

**Comment by isbluis** [comment]
This is preexisting logging code (yes, already in CI) that writes to `stdout`:

![image](https://github.com/RTXteam/RTX/assets/37852256/4ea0af94-d896-4380-8a2d-d2d0a268978f)

I could update it to go to `stderr` instead and add a bit more diagnostic logging to help trace this.

[ Also, when not running in test mode, timing is under 0.2sec -- which seems consistent with what we are observing in the wild ]

**Comment by edeutsch** [comment]
> This is preexisting logging code (yes, already in CI) that writes to `stdout`:
> 
> ![image](https://user-images.githubusercontent.com/37852256/260100837-4ea0af94-d896-4380-8a2d-d2d0a268978f.png)
> 
> I could update it to go to `stderr` instead and add a bit more diagnostic logging to help trace this.

yes, please do!

> [ Also, when not running in test mode, timing is under 0.2sec -- which seems consistent with what we are observing in the wild ]

**Comment by saramsey** [comment]
I concur with Eric-- moving the S3 bucket from `us-west-2` to `us-east-1` (I think that's where ITRB is hosted?) is a very good _post-September_ system enhancement. My feeling is that we will have our hands very full getting issues 2045 and 2022 closed out and deployed into ITRB TEST (along with a deployment of KG2.8.4c) by COB Friday.

**Comment by edeutsch** [comment]
So any chance of getting this issue fixed before the code freeze?

**Comment by isbluis** [comment]
> So any chance of getting this issue fixed before the code freeze?

I'll try adding more debug code, since none has helped remote-test that ITRB instance.

I did notice this quite a bit on the logs -- araxconfig key keeps getting "permanently added" dozens of times for this single query.  Perhaps a clue?  (I had not noticed this before.)

![image](https://github.com/RTXteam/RTX/assets/37852256/232bef15-a48f-4218-bc2c-c8de963c08ed)

I can look more into it later, but perhaps you or @saramsey know this part of the system better?

**Comment by edeutsch** [comment]
I do not know this part myself, maybe @saramsey. My hunch is that we are running RTXConfig over and over again, because many parts of the system need information from RTXConfig, and passing it around is very awkward. So ideally RTXConfig would only do slow things once and then cache all that information into a class variable so that all subsequent calls have that information already available with no slow work.

I'm not certain that would lead to a large speed increase, but it certain would feel much less wasteful.

@isbluis Whatever you can do to speed up KG2 is very helpful. It is much slower on ITRB instances now and for the September rollout, that's where it counts.

**Comment by isbluis** [comment]
One more find with the new logging code:  it looks like the pickle file is not being generated or stored.  This costs about 3.5 sec, since it has to download a YAML file and parse it on the fly.  For each query.

![image](https://github.com/RTXteam/RTX/assets/37852256/81eac753-7497-4b2e-a458-b11ccb2ab777)

Is there a procedure to pre-generate this file (and it's side-by-side JSON version)?  Not sure whom to tag, since Amy is away...

**Comment by isbluis** [comment]
Further observations when POSTing directly via curl:
`time curl --data @example1.json --header "Content-Type: application/json" https://kg2.ci.transltr.io/api/rtxkg2/v1.4/query > response.json`

**1. Mismatch in reported processing time vs. total fetch time**
rtxkg2 self-reports a total time of 2.77sec, but the total time via curl is 2 seconds longer:
![image](https://github.com/RTXteam/RTX/assets/37852256/b94b83fe-d048-4245-8d95-29ebf0a40b38)
![image](https://github.com/RTXteam/RTX/assets/37852256/07dfc16f-f1c2-4790-904d-97906663cf25)

**2. First time sink (1.7sec) happens while creating a template response**
Most likely related to YAML parsing described in comment above
![image](https://github.com/RTXteam/RTX/assets/37852256/aeb4c9a5-37da-485f-9b7f-4c9f69c957f7)

**3. Second delay (0.9sec) is in connecting to the ResponseCache**
![image](https://github.com/RTXteam/RTX/assets/37852256/ddf37b7c-98df-439a-a90f-e76b601afcbf)

All told, those two items account for 2.6 out of 2.77 seconds (94%) that rtxkg2 spends handing this query.

**Comment by isbluis** [comment]
By way of comparison, posting the same to beta one gets a response in 0.7sec:

`time curl -k --data @example1.json --header "Content-Type: application/json" https://arax.ncats.io/beta/api/rtxkg2/v1.4/query > response_beta.json`

**1. Mismatch is 0.2 seconds:**
![image](https://github.com/RTXteam/RTX/assets/37852256/2ba8f7b8-ad86-446d-a9fc-d5b1db5e583b)
![image](https://github.com/RTXteam/RTX/assets/37852256/84907c14-6759-408d-bcbc-62d77cb86e07)

**2. Creation of template response takes <0.1 sec (presumably not parsing YAML?)**
![image](https://github.com/RTXteam/RTX/assets/37852256/01b71360-748e-4b3c-9fad-55284094923f)

**3. Connecting to ResponseCache takes <0.02 sec:**
![image](https://github.com/RTXteam/RTX/assets/37852256/a38e94bb-3e6e-42f9-9c76-6f4efa01552c)

**Comment by edeutsch** [comment]
Great work, thanks!

Yes, Amy is away, so you're the best expert we have now! Can you fix it?

The missing extra time is probably just Internet latency and transmission time, not much we can do about that. It's much faster to arax.ncats.io than ITRB from Seattle because it's in an Oregon data center. Connections to ITRB pay an extra overhead from here while testing. But the reverse is likely true when communicating with other ITRB instances, which is what counts. So I would just focus on the Biolink YAML problem. If you can fix that, that will help quite a bit.

**Comment by isbluis** [comment]
Of note, that "missing extra time" is also present when initiating the query via the CI/ITRB interface:

![image](https://github.com/RTXteam/RTX/assets/37852256/5e92afba-082c-460e-ae3c-ebc706a399ff)


The rtxkg2 logs (on CI) show:

![image](https://github.com/RTXteam/RTX/assets/37852256/2d2c41bb-47cf-4af6-a6c0-c669374629b0)
![image](https://github.com/RTXteam/RTX/assets/37852256/e233284f-340a-4fac-a55c-0d0447488e32)

---
So about 2.6sec vs 4.6sec -- or **about 2sec "lost"**.

**Comment by isbluis** [comment]
New debug info rules out Biolink, and now seems to point to QueryTracker.  Adding more debug therein...

![image](https://github.com/RTXteam/RTX/assets/37852256/8ce614cc-bbf8-460b-a074-c4863ee55548)

**Comment by edeutsch** [comment]
ah, the plot thickens..

**Comment by saramsey** [comment]
Thank you for your detective work, @isbluis !

**Comment by isbluis** [comment]
New clues: **ARAXQueryTracker** spends over one second creating a session; followed by another 0.6sec wasted elsewhere (even more debug code forthcoming...)

![image](https://github.com/RTXteam/RTX/assets/37852256/e0907ad1-37bb-40b2-8156-d22e40c78669)

The relevant part of the code:
![image](https://github.com/RTXteam/RTX/assets/37852256/6275762c-37a5-4d9d-a3a8-91721641758d)

**Comment by edeutsch** [comment]
It may be that the penalty here is data centers again. I see this on arax.ncats.io/beta:
```
2023-08-18T05:52:47.581586: DEBUG: In ARAXQueryTracker init
2023-08-18T05:52:47.585029: DEBUG: ARAXQueryTracker initiating DB connection
2023-08-18T05:52:47.609782: DEBUG: ARAXQueryTracker establishing session
2023-08-18T05:52:47.628163: DEBUG: ARAXQueryTracker initialized
```
So 0.05 seonds

but on arax.ci:
```
2023-08-18T05:08:51.183315: DEBUG: In ARAXQueryTracker init
2023-08-18T05:08:51.187591: DEBUG: ARAXQueryTracker initiating DB connection
2023-08-18T05:08:51.208555: DEBUG: ARAXQueryTracker establishing session
2023-08-18T05:08:52.256380: DEBUG: ARAXQueryTracker initialized
```

1.1 seconds.

My conclusion is that which data center the MySQL server is in makes a big difference. 0.05 seconds vs 1.1 seconds difference.

Maybe after the code freeze we should consider moving the MySQL server and the S3 bucket into the same data center at the ITRB computers. ncats.io will then be paying the penalty, but that is better I think. I think those moves don't require and code changes? Just some moving of data and DNS magic?

**Comment by isbluis** [comment]
Indeed, and this adds up as the query_tracker is called multiple times.

On top of the ~1sec to initialize, it takes another ~0.7sec to create the query records:

![image](https://github.com/RTXteam/RTX/assets/37852256/2354ae52-b0ab-4452-b2e7-d5b804367854)

I believe those "lost 2 seconds" at the end of the queries come from deleting the OngoingQuery:

![image](https://github.com/RTXteam/RTX/assets/37852256/c54272e5-5aec-4b2f-bb33-1dc6e3dd9b37)

(1 sec to initialize, and likely another to execute the delete)

**Both of these add ~3.7sec (1.7+2.0) to the processing time.**

Moving the DB host sounds reasonable.

**Comment by saramsey** [comment]
Yes, there is an RTX issue on moving `arax-responses.rtx.ai` to a different AWS region:  https://github.com/RTXteam/RTX/issues/2097

**Comment by isbluis** [comment]
Are we happy with the ITRB speeds now?  Or should we keep this issue open?

**Comment by edeutsch** [comment]
yeah, let's close it. I think it would be good to do another round of optimization after we have everything deployed in the east. Maybe that can be one of our goals for the next period.

**Comment by saramsey** [comment]
Well done. This issue thread was very helpful for the progress report.
